kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kms.service;
+
+import java.security.MessageDigest;
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Objects;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.crypto.SecretKey;
+import javax.security.auth.DestroyFailedException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * A SecretKey that has RAW encoding and can be {@linkplain #destroy() destroyed}
+ * (unlike {@link javax.crypto.spec.SecretKeySpec}).
+ */
+@NotThreadSafe
+public final class DestroyableRawSecretKey implements SecretKey {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DestroyableRawSecretKey.class);
+
+    private final String algorithm;
+    private boolean destroyed = false;
+    private final byte[] key;
+
+    private DestroyableRawSecretKey(String algorithm, byte[] bytes) {
+        this.algorithm = Objects.requireNonNull(algorithm).toLowerCase(Locale.ROOT);
+        this.key = Objects.requireNonNull(bytes);
+    }
+
+    /**
+     * Create a new key by becoming owner of the given key material.
+     * The caller should not modify the given bytes after calling this method.
+     * @param algorithm The key algorithm
+     * @param bytes The key material
+     * @return The new key
+     */
+    public static @NonNull DestroyableRawSecretKey byOwnershipTransfer(@NonNull String algorithm, @NonNull byte[] bytes) {
+        return new DestroyableRawSecretKey(algorithm, bytes);
+    }
+
+    /**
+     * Create a new key by creating a copy of the given key material.
+     * @param algorithm The key algorithm
+     * @param bytes The key material
+     * @return The new key
+     */
+    public static @NonNull DestroyableRawSecretKey byClone(@NonNull String algorithm, @NonNull byte[] bytes) {
+        return new DestroyableRawSecretKey(algorithm, Objects.requireNonNull(bytes).clone());
+    }
+
+    /**
+     * Convert the given key to a destroyable key, attempting to destroy the given key.
+     * @param source The key to convert
+     * @return The new destroyable key.
+     */
+    public static @NonNull DestroyableRawSecretKey toDestroyableKey(@NonNull SecretKey source) {
+        Objects.requireNonNull(source);
+        if (!"RAW".equalsIgnoreCase(source.getFormat())) {
+            throw new IllegalArgumentException("RAW-format key required");
+        }
+        // no need to copy, because getEncoded should itself return a copy of the key material
+        var result = byOwnershipTransfer(source.getAlgorithm(), source.getEncoded());
+        try {
+            source.destroy();
+        }
+        catch (DestroyFailedException e) {
+            LOGGER.warn("Failed to destroy key of {}", source.getClass());
```

Comment:
Should we log the destroy failed exception? Or is that effectively a proxy for unsupported operation?

Reply:
We can't throw if we want this method to work with `SecretKeySpec`, which is returned by other parts of JCE and which 3rd party implementers of `Kms` are likely to be using. And if we do throw, we are in a similar situation to when we were using DRSK in the `Kms` interface, i.e, you _have_ to be using a `SecreyKey` that is actually destroyable or it won't work. Only doing it this way means it's a runtime exception, rather than an compile time error.

If that's our requirement I'd prefer we reflected it in a type. We could do that, I suppose, by subclassing `SecretKey` to remove the `throws DestroyFailedException`.

Reply:
To be clear I wasn't objecting to what it currently does, I was just wondering if we should include the exception we caught in the logging.